File: backend/services/rules_service.py
from flask_smorest import Blueprint
from flask import session, request
from sqlalchemy.exc import IntegrityError
from models.user.user import User, UserRole, db
from models.group.group import Group
from models.activity.activity import Activity
from models.rules.rules import RuleTemplate, RuleType, group_rules, activity_rules
from utils.decorators import login_required
from marshmallow import Schema, fields, validate, ValidationError

blp = Blueprint("Rules", "rules", url_prefix="/api/rules", description="Rules management routes")

class AttachRulesSchema(Schema):
    template_ids = fields.List(fields.Int(), required=True, validate=validate.Length(min=1))

class RulesService:
    """Service for managing rules templates and assignments"""
    
    @staticmethod
    def get_templates_for_type(rule_type):
        """Get all active rule templates for a specific type"""
        return RuleTemplate.get_templates_for_type(RuleType(rule_type))
    
    @staticmethod
    def attach_group_rules(group_id, template_ids, user_id):
        """Attach rule templates to a group"""
        group = Group.query.get(group_id)
        if not group:
            raise ValueError("Group not found")
        
        # Check if user can modify group rules (creator, organizer, or admin)
        user = User.query.get(user_id)
        if not (user.is_organizer_or_admin() or group.created_by == user_id):
            raise ValueError("Not authorized to modify group rules")
        
        # Remove existing rules
        db.session.execute(group_rules.delete().where(group_rules.c.group_id == group_id))
        
        # Add new rules
        for template_id in template_ids:
            template = RuleTemplate.query.get(template_id)
            if template and template.is_active and template.rule_type in [RuleType.GROUP, RuleType.BOTH]:
                db.session.execute(
                    group_rules.insert().values(
                        group_id=group_id,
                        rule_template_id=template_id
                    )
                )
        
        db.session.commit()
        return True
    
    @staticmethod
    def attach_activity_rules(activity_id, template_ids, user_id):
        """Attach rule templates to an activity"""
        activity = Activity.query.get(activity_id)
        if not activity:
            raise ValueError("Activity not found")
        
        # Check if user can modify activity rules (creator, organizer, or admin)
        user = User.query.get(user_id)
        if not (user.is_organizer_or_admin() or activity.created_by == user_id):
            raise ValueError("Not authorized to modify activity rules")
        
        # Remove existing rules
        db.session.execute(activity_rules.delete().where(activity_rules.c.activity_id == activity_id))
        
        # Add new rules
        for template_id in template_ids:
            template = RuleTemplate.query.get(template_id)
            if template and template.is_active and template.rule_type in [RuleType.ACTIVITY, RuleType.BOTH]:
                db.session.execute(
                    activity_rules.insert().values(
                        activity_id=activity_id,
                        rule_template_id=template_id
                    )
                )
        
        db.session.commit()
        return True
    
    @staticmethod
    def get_group_rules(group_id):
        """Get all rules for a group"""
        rules_query = db.session.query(RuleTemplate).join(
            group_rules, RuleTemplate.id == group_rules.c.rule_template_id
        ).filter(group_rules.c.group_id == group_id)
        
        return rules_query.all()
    
    @staticmethod
    def get_activity_rules(activity_id):
        """Get all rules for an activity"""
        rules_query = db.session.query(RuleTemplate).join(
            activity_rules, RuleTemplate.id == activity_rules.c.rule_template_id
        ).filter(activity_rules.c.activity_id == activity_id)
        
        return rules_query.all()

# REST endpoints
@blp.route("/templates", methods=["GET"])
def get_rule_templates():
    """Get rule templates filtered by type"""
    rule_type = request.args.get('type', 'BOTH')
    
    try:
        templates = RulesService.get_templates_for_type(rule_type)
        return {
            'templates': [template.to_dict() for template in templates]
        }
    except ValueError as e:
        blp.abort(400, message=str(e))

@blp.route("/groups/<int:group_id>/rules", methods=["POST"])
@blp.arguments(AttachRulesSchema)
@login_required
def attach_group_rules(group_id, args):
    """Attach rules to a group"""
    user_id = session.get('user_id')
    
    try:
        RulesService.attach_group_rules(group_id, args['template_ids'], user_id)
        return {'message': 'Rules attached to group successfully'}
    except ValueError as e:
        blp.abort(400, message=str(e))
    except Exception as e:
        blp.abort(500, message="Failed to attach rules to group")

@blp.route("/groups/<int:group_id>/rules", methods=["GET"])
def get_group_rules(group_id):
    """Get rules for a group"""
    try:
        rules = RulesService.get_group_rules(group_id)
        return {
            'group_id': group_id,
            'rules': [rule.to_dict() for rule in rules]
        }
    except Exception as e:
        blp.abort(500, message="Failed to get group rules")

@blp.route("/activities/<int:activity_id>/rules", methods=["POST"])
@blp.arguments(AttachRulesSchema)
@login_required
def attach_activity_rules(activity_id, args):
    """Attach rules to an activity"""
    user_id = session.get('user_id')
    
    try:
        RulesService.attach_activity_rules(activity_id, args['template_ids'], user_id)
        return {'message': 'Rules attached to activity successfully'}
    except ValueError as e:
        blp.abort(400, message=str(e))
    except Exception as e:
        blp.abort(500, message="Failed to attach rules to activity")

@blp.route("/activities/<int:activity_id>/rules", methods=["GET"])
def get_activity_rules(activity_id):
    """Get rules for an activity"""
    try:
        rules = RulesService.get_activity_rules(activity_id)
        return {
            'activity_id': activity_id,
            'rules': [rule.to_dict() for rule in rules]
        }
    except Exception as e:
        blp.abort(500, message="Failed to get activity rules")

File: frontend/activamigos-frontend/src/app/core/services/rules.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from 'src/environments/environment';

export interface RuleTemplate {
  id: number;
  title: string;
  description: string;
  icon: string;
  rule_type: 'GROUP' | 'ACTIVITY' | 'BOTH';
  is_active: boolean;
  created_at: string;
}

export interface RuleTemplatesResponse {
  templates: RuleTemplate[];
}

export interface RulesResponse {
  group_id?: number;
  activity_id?: number;
  rules: RuleTemplate[];
}

@Injectable({
  providedIn: 'root'
})
export class RulesService {
  private readonly API_BASE_URL = `${environment.apiUrl}/api`;

  constructor(private http: HttpClient) {}

  /**
   * Get rule templates filtered by type
   */
  getRuleTemplates(type: 'GROUP' | 'ACTIVITY' | 'BOTH' = 'BOTH'): Observable<RuleTemplatesResponse> {
    const params = { type };
    
    return this.http.get<RuleTemplatesResponse>(`${this.API_BASE_URL}/rules/templates`, {
      params,
      withCredentials: true
    }).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Attach rules to a group
   */
  attachGroupRules(groupId: number, templateIds: number[]): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(
      `${this.API_BASE_URL}/rules/groups/${groupId}/rules`,
      { template_ids: templateIds },
      { withCredentials: true }
    ).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Get rules for a group
   */
  getGroupRules(groupId: number): Observable<RulesResponse> {
    return this.http.get<RulesResponse>(
      `${this.API_BASE_URL}/rules/groups/${groupId}/rules`,
      { withCredentials: true }
    ).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Attach rules to an activity
   */
  attachActivityRules(activityId: number, templateIds: number[]): Observable<{ message: string }> {
    return this.http.post<{ message: string }>(
      `${this.API_BASE_URL}/rules/activities/${activityId}/rules`,
      { template_ids: templateIds },
      { withCredentials: true }
    ).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Get rules for an activity
   */
  getActivityRules(activityId: number): Observable<RulesResponse> {
    return this.http.get<RulesResponse>(
      `${this.API_BASE_URL}/rules/activities/${activityId}/rules`,
      { withCredentials: true }
    ).pipe(
      catchError(this.handleError)
    );
  }

  /**
   * Get rules for a specific context (group or activity)
   */
  getContextRules(contextType: 'GROUP' | 'ACTIVITY', contextId: number): Observable<RulesResponse> {
    if (contextType === 'GROUP') {
      return this.getGroupRules(contextId);
    } else {
      return this.getActivityRules(contextId);
    }
  }

  /**
   * Attach rules to a specific context (group or activity)
   */
  attachContextRules(contextType: 'GROUP' | 'ACTIVITY', contextId: number, templateIds: number[]): Observable<{ message: string }> {
    if (contextType === 'GROUP') {
      return this.attachGroupRules(contextId, templateIds);
    } else {
      return this.attachActivityRules(contextId, templateIds);
    }
  }

  /**
   * Handle HTTP errors
   */
  private handleError(error: any): Observable<never> {
    console.error('Rules service error:', error);
    throw error;
  }
}